sql/table.cc

```diff
@@ -8479,6 +8484,199 @@ LEX_CSTRING *fk_option_name(enum_fk_option opt)
+  init_one_table(LEX_STRING_WITH_LEN(MYSQL_SCHEMA_NAME),
+                 XSTRING_WITH_LEN(table_name), table_name, rw ? TL_WRITE : TL_READ);
+  open_tables_backup= new Open_tables_backup;
+  if (open_tables_backup)
+    open_log_table(thd, this, open_tables_backup);
+  else
+    my_error(ER_OUT_OF_RESOURCES, MYF(0));
+}
+
+TR_table::~TR_table()
+{
+  if (table)
+    close_log_table(thd, open_tables_backup);
+  delete open_tables_backup;
+}
+
+void TR_table::store(uint field_id, ulonglong val)
+{
+  table->field[field_id]->store(val, true);
+  table->field[field_id]->set_notnull();
+}
+
+void TR_table::store(uint field_id, timeval ts)
+{
+  table->field[field_id]->store_timestamp(ts.tv_sec, ts.tv_usec);
+  table->field[field_id]->set_notnull();
+}
+
+void TR_table::store_data(ulonglong trx_id, ulonglong commit_id, timeval &commit_ts)
+{
+  timeval start_time= {thd->start_time, thd->start_time_sec_part};
+  store(FLD_TRX_ID, trx_id);
+  store(FLD_COMMIT_ID, commit_id);
+  store(FLD_COMMIT_TS, commit_ts);
+  store(FLD_BEGIN_TS, start_time);
+  store(FLD_ISO_LEVEL, thd->tx_isolation);
+}
+
+bool TR_table::update()
+{
+  if (!table)
+    return true;
+
+  DBUG_ASSERT(table->s);
+  handlerton *hton= table->s->db_type();
+  DBUG_ASSERT(hton);
+  DBUG_ASSERT(hton->flags & HTON_NATIVE_SYS_VERSIONING);
+
+  if (hton->vers_get_trt_data(*this))
+  {
+    int error= table->file->ha_write_row(table->record[0]);
+    if (error)
+    {
+      table->file->print_error(error, MYF(0));
+    }
+    return error;
+  }
+  return false;
+}
+
+#define newx new (thd->mem_root)
+bool TR_table::query(ulonglong trx_id)
+{
+  if (!table)
+    return false;
+  SQL_SELECT_auto select;
+  READ_RECORD info;
+  int error;
+  List<TABLE_LIST> dummy;
+  SELECT_LEX &slex= thd->lex->select_lex;
+  Name_resolution_context_backup backup(slex.context, *this);
+  Item *field= newx Item_field(thd, &slex.context, (*this)[FLD_TRX_ID]);
+  Item *value= newx Item_int(thd, trx_id);
+  COND *conds= newx Item_func_eq(thd, field, value);
+  if ((error= setup_conds(thd, this, dummy, &conds)))
+    return false;
+  select= make_select(table, 0, 0, conds, NULL, 0, &error);
+  if (error || !select)
+    return false;
+  // FIXME: (performance) force index 'transaction_id'
+  error= init_read_record(&info, thd, table, select, NULL,
+                          1 /* use_record_cache */, true /* print_error */,
+                          false /* disable_rr_cache */);
+  while (!(error= info.read_record()) && !thd->killed && !thd->is_error())
+  {
+    if (select->skip_record(thd) > 0)
+      return true;
+  }
+  return false;
+}
+
+bool TR_table::query(MYSQL_TIME &commit_time, bool backwards)
+{
+  if (!table)
+    return false;
+  SQL_SELECT_auto select;
+  READ_RECORD info;
+  int error;
+  List<TABLE_LIST> dummy;
+  SELECT_LEX &slex= thd->lex->select_lex;
+  Name_resolution_context_backup backup(slex.context, *this);
+  Item *field= newx Item_field(thd, &slex.context, (*this)[FLD_COMMIT_TS]);
```

Comment:
Allocations on every examined row. This memory won't be freed until the query end, right? So such query on a big table will eat too much memory. I don't think this code scales well. We need to create `READ_RECORD` and surrounding stuff once per query and reuse. Let it be future work. For now I suggest to add a comment or open and issue on that topic.

Reply:
Makes sense. Created #314